daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -960,16 +970,67 @@ func (devices *DeviceSet) closeTransaction() error {
 	return nil
 }
 
+func determineDriverCapabilities(version string) error {
+	/*
+	 * Driver version 4.27.0 and greater support deferred activation
+	 * feature.
+	 */
+
+	logrus.Debugf("devicemapper: driver version is %s", version)
+
+	versionSplit := strings.Split(version, ".")
+	major, err := strconv.Atoi(versionSplit[0])
+	if err != nil {
+		return graphdriver.ErrNotSupported
+	}
+
+	if major > 4 {
+		DriverDeferredRemovalSupport = true
+		return nil
+	}
+
+	if major < 4 {
+		return nil
+	}
+
+	minor, err := strconv.Atoi(versionSplit[1])
+	if err != nil {
+		return graphdriver.ErrNotSupported
+	}
+
+	/*
+	 * If major is 4 and minor is 27, then there is no need to
+	 * check for patch level as it can not be less than 0.
+	 */
+	if minor >= 27 {
+		DriverDeferredRemovalSupport = true
+		return nil
+	}
+
+	return nil
+}
+
 func (devices *DeviceSet) initDevmapper(doInit bool) error {
 	// give ourselves to libdm as a log handler
 	devicemapper.LogInit(devices)
 
-	_, err := devicemapper.GetDriverVersion()
+	version, err := devicemapper.GetDriverVersion()
 	if err != nil {
 		// Can't even get driver version, assume not supported
 		return graphdriver.ErrNotSupported
 	}
 
+	if err := determineDriverCapabilities(version); err != nil {
+		return graphdriver.ErrNotSupported
+	}
+
+	// If user asked for deferred removal and both library and driver
+	// supports deferred removal use it.
+	if EnableDeferredRemoval && DriverDeferredRemovalSupport && devicemapper.LibraryDeferredRemovalSupport == true {
```

Comment:
Ha. That is so explicit that it did not stick out to me.
On May 4, 2015 12:55, "Brian Goff" <notifications@github.com> wrote:

> In daemon/graphdriver/devmapper/deviceset.go
> <https://github.com/docker/docker/pull/12190#discussion_r29616949>:
>
> >  	if err != nil {
> >  		// Can't even get driver version, assume not supported
> >  		return graphdriver.ErrNotSupported
> >  	}
> >
> > +	if err := determineDriverCapabilities(version); err != nil {
> > +		return graphdriver.ErrNotSupported
> > +	}
> > +
> > +	// If user asked for deferred removal and both library and driver
> > +	// supports deferred removal use it.
> > +	if EnableDeferredRemoval && DriverDeferredRemovalSupport && devicemapper.LibraryDeferredRemovalSupport == true {
>
> == true is not needed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/12190/files#r29616949>.
>